php/php-rr-sdk.md

```diff
@@ -0,0 +1,692 @@
+
+yield new WaitAll($a, $b);
+
+// or 
+yield new WaitAny($a, $b);
+```
+
+> More atomic blocks can be added down the road.
+
+#### Syntax Sugar
+It should be possible to mock `new Activity` and other commands with shorted methods from abstract workflow implementation:
+
+```php
+class MyWorkflow extends Workflow 
+{
+    public function run()
+    {
+        $result = yield $this->activities->doSomething()->get();
+        yield $this->sleep(1);
+
+        // ...
+    }
+}
+``` 
+
+Examples below will be given using sugared approach for simplicity (it is possible to fallback to commands as well). 
+
+#### Timers
+Workflow timers can operate as promise or blocking operation:
+
+```php
+$timer = yield $this->timer(1 * Time::DAY);
+
+yield $timer->wait();
+```
+
+Blocking:
+
+```php
+yield $this->timer(1 * Time::DAY)->wait();
+
+// or
+yield $this->sleep(1 * Time::DAY);
+```
+
+Can be comfined with Promises:
+
+```php
+$a = yield $this->activities->doSomething(...);
+$t = yield $this->timer(1 * Time::MINUTE);
+
+yield $this->waitAny($a, $t);
+
+if ($t->isFired()) {
+    // ...
+}
+```
+
+### Queries
+Workflow queries can be easily implemented using dedicated query method of workflow. The state of the workflow can be 
+extracted from the properties of the object.
+
+```php
+use Temporal\Workflow;
+
+class DemoWorkflow extends Workflow\Workflow
+{
+    private $step = 0;
+
+    public function queryStep(): int 
+    {
+        return $this->step;
+    }
+
+    public function run(string $input): string
+    {
+        yield $this->activities->doSomething($input);
+        $this->step++;
+
+        // ...
+    }
+}
+```
+
+> Query methods must be registered explicitly.
+
+An alternative approach can be based on registering the callback function during the workflow execution:
+
+```php
+yield new QueryHandler('something', function() use (&$step) {
+    return $step;
+});
+```
+
+> Such an approach is more suitable for dynamic workflows. The principle of work is identical to pre-registered queries.
+
+#### Using Annotations
+For more advanced frameworks it should be possible to register query method using annotation or attribute (PHP8) as 
+in Java SDK.
+
+```php
+/**
+ * @Workflow\QueryMethod(name="something")
+ */
+public function querySomething(string $world): string 
+{
+    return 'hello';
+}
+```
+
+### Signals
+Unlike queries the signals are fist-class citizen of actual workflow code. Unlike activities the signal can be triggered
+at any moment, without prior expectation by workflow. 
+
+Signal method can be defined explicitly:
+
+```php
+class UploadWorkflow extends Workflow\Workflow
+{
+    /** @Temporal\SignalMethod(name="signal") */
+    public function handleSignal($input):void
+    {
+    
+    }
+
+    public function run(string $input): string
+    {
+        
+    }
+}
+```
+
+Or at runtime:
+
+```php
+public function run(string $input): string
+{
+    yield new SignalHandler('name', function($input){
+    
+    });       
+}
+```
```

Comment:
Open question is what happens if you register more than one handler for the given signal type. Will all of them be called? Or double registration is rejected?

Reply:
That's a good question... practically I'm seeing now issues handling query in multiple handlers. Conceptually it might cause some side effects. However, it can also be used to build some more complex things...